File: cpp/src/2020/aug/reorder-list.cpp
#include <common>

using namespace common

// O(N) time, O(1) space in total
void reorderList(ListNode *head) {
    if (!head || !head->next) return;

    // find the middle node: O(n)
    ListNode *p1 = head, *p2 = head->next;
    while (p2 && p2->next) {
        p1 = p1->next;
        p2 = p2->next->next;
    }

    // cut from the middle and reverse the second half: O(n)
    ListNode *head2 = p1->next;
    p1->next = NULL;

    p2 = head2->next;
    head2->next = NULL;
    while (p2) {
        p1 = p2->next;
        p2->next = head2;
        head2 = p2;
        p2 = p1;
    }

    // merge two lists: O(n)
    for (p1 = head, p2 = head2; p1;) {
        auto t = p1->next;
        p1 = p1->next = p2;
        p2 = t;
    }

    //for (p1 = head, p2 = head2; p2; ) {
    //    auto t = p1->next;
    //    p1->next = p2;
    //    p2 = p2->next;
    //    p1 = p1->next->next = t;
    //}
}

class Solution3 {
public:
    ListNode *reorderList(ListNode *head, int len) {
        if (len == 0)
            return NULL;
        if (len == 1)
            return head;
        if (len == 2)
            return head->next;
        ListNode *tail = reorderList(head->next, len - 2);
        ListNode *tmp = tail->next;  
        tail->next = tail->next->next;
        tmp->next = head->next;
        head->next = tmp;
        return tail;
    }

    void reorderList(ListNode *head) {  //recursive
        ListNode *tail = NULL;
        tail = reorderList(head, getLength(head));
    }
};

File: code/java/src/main/java/cn/az/code/year2020/aug/ReorderList.java
package cn.az.code.year2020.aug;

import cn.az.code.common.ListNode;

/**
 * @author az
 * @since 08/20/20
 */
public class ReorderList {

    public static void main(String[] args) {
        ReorderList rl = new ReorderList();
        ListNode root = ListNode.forSimpleTest(1, 2, 3, 4, 5, 6);
        System.out.println(root);
        rl.reorderList(root);

    }

    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }

        // step 1. cut the list to two halves
        // prev will be the tail of 1st half
        // slow will be the head of 2nd half
        ListNode prev = null, slow = head, fast = head;

        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }

        prev.next = null;

        // step 2. reverse the 2nd half
        ListNode l2 = reverse(slow);

        // step 3. merge the two halves
        merge(head, l2);
    }

    ListNode reverse(ListNode head) {
        ListNode prev = null, curr = head, next;

        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }

    void merge(ListNode l1, ListNode l2) {
        while (l1 != null) {
            ListNode n1 = l1.next, n2 = l2.next;
            l1.next = l2;

            if (n1 == null) {
                break;
            }

            l2.next = n1;
            l1 = n1;
            l2 = n2;
        }
    }
}
